### PR TITLE
error processing refactor

### PR DESCRIPTION
**Pull request by matei-p**
## Description of the reasoning behind the pull request (what feature was missing / how the problem was manifesting itself / what was the motive behind the refactoring)
Existing tx processing failure mechanism is encompassed in ProcessIfError(). This code was scattered everywhere and there was also the risk of double calling it.
  
## Proposed Changes
Refactor of the failure processing calling code, by moving it as high as possible in the calling hierarchy.

## Testing procedure
Existing unit tests cover the error scenarios and their pass guarantees a correct refactor.

### CODE REVIEW

**Review by sasurobert — COMMENTED** [review]
I like the idea - resolve TODOs

File: process/smartContract/processorV2/failureContext.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	processFail   bool
+	txHash        []byte
+	errorMessage  string
+	returnMessage []byte
+	snapshot      int
+	gasLocked     uint64
+	logs          []*vmcommon.LogEntry
+}
+
+func NewFailureContext(sc *scProcessor) *failureContext {
```
Comment (iulianpascalau) [inline process/smartContract/processorV2/failureContext.go:16]: providing the sc processor pointer here suggest that the failureContext and scProcessor are closely tight together.
In this case, I suggest making the scProcessorV2 keep a pointer of type failure context inside it. In this way you will not need to pass it around on all methods. You will need to reset it whenever the sc processor needs to start processing something.

File: process/smartContract/processorV2/failureContext.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+		processFail: false,
+	}
+}
+
+func (failCtx *failureContext) makeSnaphot() *failureContext {
+	failCtx.snapshot = failCtx.sc.accounts.JournalLen()
+	return failCtx
+}
+
+func (failCtx *failureContext) setTxHash(txHash []byte) *failureContext {
```
Comment (iulianpascalau) [inline process/smartContract/processorV2/failureContext.go:28]: daisy chaining that does not quite add clarity but makes the calls a one-liner.
I always prefer multiple calls on multiple lines as long as they are clear enough.

File: process/smartContract/processorV2/processV2.go [hunk 3]
```diff
@@ -825,27 +846,50 @@ func (sc *scProcessor) doExecuteBuiltInFunction(
...
+	acntSnd, acntDst state.UserAccountHandler,
+	failureContext *failureContext,
+) (vmcommon.ReturnCode, error) {
+	returnCode, vmInput, txHash, err := sc.prepareExecution(tx, acntSnd, acntDst, true, failureContext)
 	if err != nil || returnCode != vmcommon.Ok {
 		return returnCode, err
 	}
 
-	snapshot := sc.accounts.JournalLen()
+	failureContext.makeSnaphot()
```
Comment (iulianpascalau) [inline process/smartContract/processorV2/processV2.go:873]: this is the second usage of these 3 functions. Suggestion to add a single function that set all these 3 values.

File: process/smartContract/processorV2/processV2.go [hunk 4]
```diff
@@ -825,27 +846,50 @@ func (sc *scProcessor) doExecuteBuiltInFunction(
...
 
 	acntSnd, err = sc.reloadLocalAccount(acntSnd)
 	if err != nil {
-		return 0, err
+		return vmcommon.Ok, err
 	}
 
 	if vmInput.ReturnCallAfterError && vmInput.CallType != vmData.AsynchronousCallBack {
-		return sc.finishSCExecution(make([]data.TransactionHandler, 0), txHash, tx, vmOutput, 0)
+		returnCode, err := sc.finishSCExecution(make([]data.TransactionHandler, 0), txHash, tx, vmOutput, 0)
```
Comment (iulianpascalau) [inline process/smartContract/processorV2/processV2.go:891]: why this change?

File: process/smartContract/processorV2/processV2.go [hunk 5]
```diff
@@ -939,20 +993,22 @@ func (sc *scProcessor) doExecuteBuiltInFunction(
...
-	if errCheck != nil {
-		return vmcommon.UserError, sc.ProcessIfError(acntSnd, txHash, tx, errCheck.Error(), []byte(errCheck.Error()), snapshot, vmInput.GasLocked)
+	err = sc.checkSCRSizeInvariant(scrResults)
+	if err != nil {
+		failureContext.setMessagesFromError(err)
+		return vmcommon.UserError, nil
 	}
 
-	return sc.finishSCExecution(scrResults, txHash, tx, newVMOutput, builtInFuncGasUsed)
+	returnCode, err = sc.finishSCExecution(scrResults, txHash, tx, newVMOutput, builtInFuncGasUsed)
```
Comment (iulianpascalau) [inline process/smartContract/processorV2/processV2.go:1010]: why these changes?

File: process/smartContract/processorV2/processV2.go (hunk 4 above)
Comment (SebastianMarian) [reply process/smartContract/processorV2/processV2.go:891]: +1

File: process/smartContract/processorV2/processV2.go [hunk 6]
```diff
@@ -939,20 +993,22 @@ func (sc *scProcessor) doExecuteBuiltInFunction(
 				tx:                   tx,
 				scrTxs:               scrResults,
 				acntSnd:              acntSnd,
-				snapshot:             snapshot,
 				createdAsyncCallback: createdAsyncCallback,
+				failureContext:       failureContext,
 			})
 		if errReturnCode != vmcommon.Ok || err != nil {
-			return errReturnCode, sc.ProcessIfError(acntSnd, txHash, tx, err.Error(), []byte(err.Error()), snapshot, vmInput.GasLocked)
+			return errReturnCode, nil
```
Comment (SebastianMarian) [inline process/smartContract/processorV2/processV2.go:1000]: failureContext.setMessagesFromError(err) before return?

File: process/smartContract/processorV2/processV2.go (hunk 5 above)
Comment (SebastianMarian) [reply process/smartContract/processorV2/processV2.go:1010]: +1

File: process/smartContract/processorV2/processV2.go (hunk 3 above)
Comment (matei-p) [reply process/smartContract/processorV2/processV2.go:873]: In my opinion would complicate a little bit the function signatures and in one of the cases we should set the gas locked only after an error checks is made so it's possible they are not linked.

File: process/smartContract/processorV2/failureContext.go (hunk 1 above)
Comment (sasurobert) [reply process/smartContract/processorV2/failureContext.go:16]: In one sense you are right, however we do not have any accumulator in this sense at SCProcessorLevel - I think it is better to keep it that way. dependency on scProcessor can be taken out if instead on makeSnapshot - we have setSnapshot(sc.accounts.journallen)

File: process/smartContract/processorV2/processV2.go [hunk 7]
```diff
@@ -899,23 +952,24 @@ func (sc *scProcessor) doExecuteBuiltInFunction(
...
 		if err != nil {
-			return vmcommon.ExecutionFailed, sc.ProcessIfError(acntSnd, txHash, tx, err.Error(), []byte(err.Error()), snapshot, vmInput.GasLocked)
+			return vmcommon.ExecutionFailed, nil
 		}
 
-		newVMOutput, errReturnCode, err = sc.executeSmartContractCallAndCheckGas(newVMInput, tx, newVMInput.CurrentTxHash, snapshot, acntSnd, newDestSC)
+		newVMOutput, errReturnCode, err = sc.executeSmartContractCallAndCheckGas(newVMInput, tx, newVMInput.CurrentTxHash, acntSnd, newDestSC, failureContext)
 		if err != nil {
-			return vmcommon.ExecutionFailed, sc.ProcessIfError(acntSnd, txHash, tx, err.Error(), []byte(err.Error()), snapshot, vmInput.GasLocked)
+			return vmcommon.ExecutionFailed, nil
```
[the remark this answers is not in the document]
Comment (matei-p) [reply process/smartContract/processorV2/processV2.go:963]: good catch

File: process/smartContract/processorV2/processV2.go (hunk 6 above)
Comment (matei-p) [reply process/smartContract/processorV2/processV2.go:1000]: good catch

File: process/smartContract/processorV2/processV2.go [hunk 8]
```diff
@@ -899,23 +952,24 @@ func (sc *scProcessor) doExecuteBuiltInFunction(
 				scExecuteOutTransfer: executionDataAfterBuiltIn.scExecuteOutTransfer,
 				tx:                   tx,
 				acntSnd:              acntSnd,
-				snapshot:             snapshot,
+				failureContext:       failureContext,
 			})
 		if err != nil {
-			return vmcommon.ExecutionFailed, sc.ProcessIfError(acntSnd, txHash, tx, err.Error(), []byte(err.Error()), snapshot, vmInput.GasLocked)
+			return vmcommon.ExecutionFailed, nil
```
[the remark this answers is not in the document]
Comment (matei-p) [reply process/smartContract/processorV2/processV2.go:958]: good catch

File: process/smartContract/processorV2/failureContext.go [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+	return failCtx
+}
+
+func (failCtx *failureContext) setMessage(errorMessage string) *failureContext {
+	failCtx.setMessages(errorMessage, []byte(errorMessage))
+	return failCtx
+}
+
+func (failCtx *failureContext) setMessagesFromError(err error) *failureContext {
+	failCtx.setMessage(err.Error())
```
[the remark this answers is not in the document]
Comment (matei-p) [reply process/smartContract/processorV2/failureContext.go:44]: no longer the case

File: process/smartContract/processorV2/failureContext.go [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+	return failCtx
+}
+
+func (failCtx *failureContext) setGasLocked(gasLocked uint64) *failureContext {
+	failCtx.gasLocked = gasLocked
+	return failCtx
+}
+
+func (failCtx *failureContext) setMessage(errorMessage string) *failureContext {
+	failCtx.setMessages(errorMessage, []byte(errorMessage))
```
[the remark this answers is not in the document]
Comment (matei-p) [reply process/smartContract/processorV2/failureContext.go:39]: no longer the case

File: process/smartContract/processorV2/processV2.go (hunk 5 above)
Comment (matei-p) [reply process/smartContract/processorV2/processV2.go:1010]: at some point there was some code between that and return :) ofc, fixed

File: process/smartContract/processorV2/processV2.go (hunk 4 above)
Comment (matei-p) [reply process/smartContract/processorV2/processV2.go:891]: at some point there was some code between that and return :) ofc, fixed

File: process/smartContract/processorV2/failureContext.go (hunk 2 above)
Comment (matei-p) [reply process/smartContract/processorV2/failureContext.go:28]: done

File: process/smartContract/processorV2/failureContext.go (hunk 1 above)
Comment (matei-p) [reply process/smartContract/processorV2/failureContext.go:16]: done

**Review by SebastianMarian — APPROVED** [review]
